### ISSUE TITLE
Incorrect handling of `filename`, `filename*` in Content-Disposition header

### ISSUE THREAD

**Issue by richdougherty** [akka/akka-http#1240]
This issue is visible in Play when we're sending a result. Play creates a header value like the following:

```
Content-Disposition: inline; filename="测 试.tmp"; filename*=utf-8''%E6%B5%8B%20%E8%AF%95.tmp
```

Akka HTTP parses this value in Play's `AkkaModelConversion` class, then renders the result as:

```
Content-Disposition: inline; filename="utf-8''%E6%B5%8B%20%E8%AF%95.tmp"
```

I believe this shows that Akka HTTP has bugs in both parsing and rendering the header.

* When parsing the header, the `filename*` parameter is preferred over the `filename`, since it comes second in the parameter list. This is correctly handled. However, the `filename*` has a special prefix (`utf8''`) that specifies the character encoding. This prefix should be used by the parser to set the character encoding, but not included in the filename itself. Akka HTTP appears to be including this value in the filename itself, rather than using it to specify a character set when decoding.

* When rendering the header, the `filename` parameter should probably only include "safe" ISO-8859-1, characters. Unsafe characters that need escaping should probably be stripped out, rather than encoded with `%`, like Akka HTTP does. This is because RFC2616 says that `\` should be used for quoting strings here. To be safe, complex characters could be excluded from the `filename` parameter and encoded properly in the `filename*` parameter instead. Modern clients will recognise and correctly handle the `filename*` parameter, ignoring the `filename` parameter. The `filename*` parameter can include the true filename, encoded and percent-escaped as per RFC6266. The suggested encoding is UTF-8, since it is mandatory for clients to support this encoding.

See:
* Related Play issue: https://github.com/playframework/playframework/issues/7501
* RFC about the filename parameter: https://tools.ietf.org/html/rfc6266#section-4.3
* Example Content-Disposition: https://tools.ietf.org/html/rfc6266#section-5
* RFC about mandatory client encodings: https://tools.ietf.org/html/rfc5987#section-3.2
* How to encode `*` parameters with non-standard characters: https://tools.ietf.org/html/rfc2231#section-4
* How to encode non-`*` parameters: https://tools.ietf.org/html/rfc2616#section-2.2

**Comment by richdougherty** [comment]
Note: more discussion of how clients handle `filename` and `filename*` parameters. May be outdated.

* https://stackoverflow.com/questions/93551/how-to-encode-the-filename-parameter-of-content-disposition-header-in-http
* http://greenbytes.de/tech/tc2231/

**Comment by jrudolph** [comment]
Thanks for the comprehensive report, @richdougherty. I agree the current behavior doesn't make sense. I discussed this before with @schmitch in #695 and also created #1066.

As I sketched in https://github.com/akka/akka-http/pull/695#issuecomment-297761574, I think this can be worked around in play by creating a RawHeader but we should also make sure that a parsing/rendering roundtrip ends up with a semantically equivalent representation.

**Comment by richdougherty** [comment]
Another reported Play issue, this time when uploading files: https://github.com/playframework/playframework/issues/7719

This ticket could be tagged with the _play-akka-http-integration_ label if you wish.

**Comment by richdougherty** [comment]
Here's a related Play issue raised by @jroper: https://github.com/playframework/playframework/issues/7733. He's reporting that the Akka HTTP backend is stripping quotes from the header parameter. While this is legal according to the RFC it would be nicer for users if Akka HTTP tried to preserve the exact formatting of a parameter when parsing it and then subsequently rendering it again.

**Comment by guntiso** [comment]
See also interoperability advice in rfc6266 Appendix D - Advice on Generating Content-Disposition Header Fields - https://tools.ietf.org/html/rfc6266#appendix-D

**Comment by raboof** [comment]
This is less relevant for Play users now since Play switched to `RawHeader`s in https://github.com/playframework/playframework/pull/8314 (2.6.13). This means we're a little more free to do what makes sense for 'direct' akka-http users, and #1066 seems like a sensible approach which AFAICS also fits the advice in that appendix.

**Comment by jrudolph** [comment]
Here's a test that fails (but also doesn't expect the completely correct behavior as suggested in the RFCs):

```
    "render utf8 filenames correctly" in {
      val formData = Multipart.FormData.Strict(
        Multipart.FormData.BodyPart.Strict("foo", "test", Map("filename" -> "中文.txt.gz")) :: Nil
      )

      formData.toEntity("aaaaa").data.utf8String shouldEqual
        """--aaaaa
          |Content-Type: text/plain; charset=UTF-8
          |Content-Disposition: form-data; filename="中文.txt.gz"; name="foo"
          |
          |test
          |--aaaaa--""".stripMargin
    }
```

**Comment by mkurz** [comment]
@jrudolph So the fix will be in upcoming 10.2.3?

**Comment by jrudolph** [comment]
> @jrudolph So the fix will be in upcoming 10.2.3?

Yes (though no release is imminent).

**Comment by arkrost** [comment]
Is 10.2.3 going to be released this year?
